cartographer/field_types/schema_attribute.py

```diff
@@ -10,49 +10,100 @@ class SchemaAttribute(object):
+        return self.sanity_check()
 
-        identifier_args = [model_attribute, serializer_method]
-        provided_identifiers = [identifier
-                                for identifier in identifier_args
-                                if identifier]
-        if len(provided_identifiers) > 1:
-            raise Exception("only one of [{}] should be provided".format(identifier_args.join(", ")))
+    def description(self, description_string):
+        """Set a description for this attribute. Useful for documentations"""
+        self.description_string = description_string
+        return self.sanity_check()
 
-        self.model_attribute = model_attribute
-        self.is_property = is_property
-        self.serializer_method = serializer_method
+    def self_explanatory(self, is_self_explanatory=True):
+        self.is_self_explanatory = is_self_explanatory
+        return self.sanity_check()
+
+    def nullable(self, is_nullable=True):
+        """Allows this property to be `null` during creation and update."""
+        self.is_nullable = is_nullable
+        return self.sanity_check()
+
+    def optional_on_create(self, is_optional_on_create=True):
+        """During creation, this property can be omitted."""
+        self.is_optional_on_create = is_optional_on_create
+        return self.sanity_check()
+
+    def computed(self, is_computed=True):
+        """This attribute can only be read and specifying it on creation or update will result in to an error."""
+        self.is_computed = is_computed
+        return self.sanity_check()
+
+    def sanity_check(self):
+        """Checks that the attribute definition is in a usable state and has no conflicts"""
+        if sum(map(bool, [self.model_property, self.model_method, self.serializer_method])) != 1:
+            raise ValueError(
+                'One of `model_property`, `model_method` or `serializer_method` should be supplied to `read_from`.',
+            )
+
+        if not self.is_self_explanatory and not self.description_string:
+            raise ValueError('Description string must be provided using `description` or marked as `self_explanatory`.')
+
+        if self.is_computed and self.is_optional_on_create:
+            raise ValueError('A `computed` property cannot be assigned during creation.')
+
+        return self
+
+    def get_value(self, serializer):
+        """
+        Extracts the attribute value using the serializer
+
+        :param serializer: The serializer which is currently responsible for serializing this value
+        :return: The raw value of the attribute before any processing
+        """
+        if self.model_property:
+            return getattr(serializer.model, self.model_property)
+        if self.model_method:
+            return getattr(serializer.model, self.model_method)()
+        if self.serializer_method:
+            return getattr(serializer, self.serializer_method)()
+
+        return None
+
+    def is_valid(self, value):
```

Comment:
I personally don't like how this method works right now since it just gives a 'yes' or 'no' where I think a better return would be a list of all things that are wrong. I'm unsure of how to do it - would strings be useful? `Error` objects?

Ideas are welcomed.